open_spiel/games/phantom_go.cc

```diff
@@ -0,0 +1,302 @@
+        return '!';
+    }
+}
+
+std::string GoState::ToString() const {
+  std::stringstream ss;
+  ss << "GoState(komi=" << komi_ << ", to_play=" << GoColorToString(to_play_)
+     << ", history.size()=" << history_.size() << ", " <<
+      "stones_count: w" << board_.stoneCount.first << " b" << board_.stoneCount.second << ")\n";
+  
+  ss << board_;
+
+
+  //update 4
+
+  ss << "\nObservation white:\n";
+
+  for (int x = board_.board_size() - 1; x >= 0; x--)
+  {
+      ss << " " << x + 1 << " ";
+      for (int y = 0; y < board_.board_size(); y++)
+      {
+          ss << GoColorToChar(board_.observationWhite[x * board_.board_size() + y]);
+      }
+      ss << "\n";
+  }
+  ss << "   ABCDEFGHJ\n";
+   
+  ss << "\nObservation black:\n";
+  for (int x = board_.board_size() - 1; x >= 0; x--)
+  {
+      ss << " " << x + 1 << " ";
+      for (int y = 0; y < board_.board_size(); y++)
+      {
+          ss << GoColorToChar(board_.observationBlack[x * board_.board_size() + y]);
+      }
+      ss << "\n";
+  }
+  ss << "   ABCDEFGHJ\n";
+
+  return ss.str();
+}
+
+bool GoState::IsTerminal() const {
+  if (history_.size() < 2) return false;
+  return (history_.size() >= max_game_length_) || superko_ ||
+         (history_[history_.size() - 1].action == board_.pass_action() &&
+          history_[history_.size() - 2].action == board_.pass_action());
+}
+
+std::vector<double> GoState::Returns() const {
+  if (!IsTerminal()) return {0.0, 0.0};
+
+  if (superko_) {
+    // Superko rules (https://senseis.xmp.net/?Superko) are complex and vary
+    // between rulesets.
+    // For simplicity and because superkos are very rare, we just treat them as
+    // a draw.
+    return {DrawUtility(), DrawUtility()};
+  }
+
+  // Score with Tromp-Taylor.
+  float black_score = TrompTaylorScore(board_, komi_, handicap_);
+
+  std::vector<double> returns(phantom_go::NumPlayers());
+  if (black_score > 0) {
+    returns[ColorToPlayer(GoColor::kBlack)] = WinUtility();
+    returns[ColorToPlayer(GoColor::kWhite)] = LossUtility();
+  } else if (black_score < 0) {
+    returns[ColorToPlayer(GoColor::kBlack)] = LossUtility();
+    returns[ColorToPlayer(GoColor::kWhite)] = WinUtility();
+  } else {
+    returns[ColorToPlayer(GoColor::kBlack)] = DrawUtility();
+    returns[ColorToPlayer(GoColor::kWhite)] = DrawUtility();
+  }
+  return returns;
+}
+
+std::unique_ptr<State> GoState::Clone() const {
+  return std::unique_ptr<State>(new GoState(*this));
+}
+
+void GoState::UndoAction(Player player, Action action) {
+  // We don't have direct undo functionality, but copying the board and
+  // replaying all actions is still pretty fast (> 1 million undos/second).
+  history_.pop_back();
+  --move_number_;
+  ResetBoard();
+  for (auto [_, action] : history_) {
+    DoApplyAction(action);
+  }
+}
+
+//need to remake
+//update 3
+void GoState::DoApplyAction(Action action) {
+  /*SPIEL_CHECK_TRUE(
```

Comment:
This invariant is commented out... why?

Reply:
I have commented this out, because board_.PlayMove() can return false, which suggests, that a plaer did not play a legal move. For the same reason, changing the player, who is on move, as on line 260, should only happen if a legal move was played. I am going to remove the commented part.